File: txmatching/configuration/configuration.py
from dataclasses import dataclass, field
from typing import List

from txmatching.configuration.subclasses import (ForbiddenCountryCombination,
                                                 ManualDonorRecipientScore,
                                                 PatientDbId)
from txmatching.utils.enums import Country

DEFAULT_FORBIDDEN_COUNTRY_LIST = [ForbiddenCountryCombination(Country.AUT, Country.IL),
                                  ForbiddenCountryCombination(Country.IL, Country.AUT)]


# pylint: disable=too-many-instance-attributes
# I think it is reasonable to have many attributes here

@dataclass
class Configuration:
    """
    Attributes:
        minimum_total_score: Minimum total score (compatibility index + blood group bonus) that is required for
        a transplant to be possible
        use_binary_scoring: If all the conditions above are satisfied, then use just 1 for possible transplant
        and -inf for impossible
    manual_donor_recipient_scores: Manual setting of score for tuple of recipient and donor
    """
    scorer_constructor_name: str = 'HLAAdditiveScorer'
    solver_constructor_name: str = 'AllSolutionsSolver'
    require_compatible_blood_group: bool = False
    minimum_total_score: float = 0.0
    maximum_total_score: float = 27.0
    require_better_match_in_compatibility_index: bool = False
    require_better_match_in_compatibility_index_or_blood_group: bool = False
    blood_group_compatibility_bonus: float = 0.0
    use_binary_scoring: bool = False
    max_cycle_length: int = 100
    max_sequence_length: int = 100
    max_number_of_distinct_countries_in_round: int = 100
    required_patient_db_ids: List[PatientDbId] = field(default_factory=list)
    use_split_resolution: bool = False
    forbidden_country_combinations: List[ForbiddenCountryCombination] = field(
        default_factory=lambda: DEFAULT_FORBIDDEN_COUNTRY_LIST)
    manual_donor_recipient_scores: List[ManualDonorRecipientScore] = field(default_factory=list)
    max_matchings_to_show_to_viewer: int = 10

File: txmatching/web/frontend/src/app/model/Configuration.ts
import { CountryCodeGenerated, ScorerGenerated, SolverGenerated, HlaCrossmatchLevelGenerated } from '@app/generated';

export interface Configuration {
  use_high_resolution: boolean;
  hla_crossmatch_level: HlaCrossmatchLevelGenerated;
  require_compatible_blood_group: boolean;
  require_better_match_in_compatibility_index: boolean;
  require_better_match_in_compatibility_index_or_blood_group: boolean;
  use_binary_scoring: boolean;

  blood_group_compatibility_bonus: number;
  minimum_total_score: number;
  max_cycle_length: number;
  max_sequence_length: number;
  max_number_of_distinct_countries_in_round: number;
  max_matchings_to_show_to_viewer: number;
  max_number_of_matchings: number;
  max_debt_for_country: number;
  max_debt_for_country_for_blood_group_zero: number;

  // complex FE
  manual_donor_recipient_scores: DonorRecipientScore[];
  required_patient_db_ids: number[];
  forbidden_country_combinations: CountryCombination[];

  scorer_constructor_name: ScorerGenerated;
  solver_constructor_name: SolverGenerated;

  // Parameters that should not be modified in FE
  max_cycles_in_all_solutions_solver: number;
  max_matchings_in_all_solutions_solver: number;
  max_matchings_in_ilp_solver: number;
  max_number_of_dynamic_constrains_ilp_solver: number;

  [key: string]: boolean | string | number | number[] | CountryCombination[] | DonorRecipientScore[];
}

export interface DonorRecipientScore {
  donor_db_id: number;
  recipient_db_id: number;
  score: number;
}

export interface CountryCombination {
  donor_country: CountryCodeGenerated;
  recipient_country: CountryCodeGenerated;
}
